kubecontrolplane/v1/types.go

```diff
@@ -0,0 +1,187 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	osinv1 "github.com/openshift/api/osin/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type KubeAPIServerConfig struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// provides the standard apiserver configuration
+	configv1.GenericAPIServerConfig `json:",inline" protobuf:"bytes,1,opt,name=genericAPIServerConfig"`
+
+	// authConfig configures authentication options in addition to the standard
+	// oauth token and client certificate authenticators
+	AuthConfig MasterAuthConfig `json:"authConfig" protobuf:"bytes,2,opt,name=authConfig"`
+
+	// aggregatorConfig has options for configuring the aggregator component of the API server.
+	AggregatorConfig AggregatorConfig `json:"aggregatorConfig" protobuf:"bytes,3,opt,name=aggregatorConfig"`
+
+	// kubeletClientInfo contains information about how to connect to kubelets
+	KubeletClientInfo KubeletConnectionInfo `json:"kubeletClientInfo" protobuf:"bytes,4,opt,name=kubeletClientInfo"`
+
+	// servicesSubnet is the subnet to use for assigning service IPs
+	ServicesSubnet string `json:"servicesSubnet" protobuf:"bytes,5,opt,name=servicesSubnet"`
+	// servicesNodePortRange is the range to use for assigning service public ports on a host.
+	ServicesNodePortRange string `json:"servicesNodePortRange" protobuf:"bytes,6,opt,name=servicesNodePortRange"`
+
+	// legacyServiceServingCertSignerCABundle is the old service serving cert signer before we switched to a separate controller
+	// TODO this should be removable in a later release after we've completed migration
+	LegacyServiceServingCertSignerCABundle string `json:"legacyServiceServingCertSignerCABundle" protobuf:"bytes,7,opt,name=legacyServiceServingCertSignerCABundle"`
+
+	// UserAgentMatchingConfig controls how API calls from *voluntarily* identifying clients will be handled.  THIS DOES NOT DEFEND AGAINST MALICIOUS CLIENTS!
+	// TODO I think we should just drop this feature.
+	UserAgentMatchingConfig UserAgentMatchingConfig `json:"userAgentMatchingConfig" protobuf:"bytes,8,opt,name=userAgentMatchingConfig"`
+
+	// imagePolicyConfig feeds the image policy admission plugin
+	// TODO make it an admission plugin config
+	ImagePolicyConfig KubeAPIServerImagePolicyConfig `json:"imagePolicyConfig" protobuf:"bytes,9,opt,name=imagePolicyConfig"`
```

Comment:
Why is image policy config so special?

Reply:
same for project config below

Reply:
protects from pulling images you don't want including the infra images?

Reply:
> Why is image policy config so special?

Instead of a regular admission plugin?  At the moment, just time.  That requires a more significant refactor. This simply strips the information down from a larger struct in openshift.